File: BinarySearch/capacityToshipPackage.py
class Solution:
    def isPossible(self, arr, D, max_capacity):
        countDays = 1
        countW = 0
        for i in arr:
            countW += i
            if i > max_capacity:
                return False
            elif countDays > D:
                return False
            elif countW > max_capacity:
                countDays += 1
                countW = i
        if countDays <= D:
            return True
        else:
            return False
        
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        total = sum(weights)
        start = weights[0]
        end = total
        result = float('inf')
        while start <= end:
            mid = start + (end - start)//2
            if (self.isPossible(weights, D, mid)):
                result = min(result, mid)
                end = mid -1
            else:
                start = mid + 1
        return result
        

File: BinarySearch/capacityToshipPackage.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool isPossible(vector<int> &weights, int D, int max_capacity){
        int countDays = 1;
        int countWeights = 0;
        for (int &i: weights){
            countWeights += i;
            if (i > max_capacity){
                return false;
            }
            else if (countDays > D){
                return false;
            }
            else if (countWeights > max_capacity){
                countDays += 1;
                countWeights = i;
            }

        }
        if (countDays <= D){
            return true;
        }
        else {
            return false;
        }
    }
    int shipWithinDays(vector<int>& weights, int D) {
        int sum = 0;
        for (int &i: weights){
            sum += i;
        }
        int start = 0;
        int end = sum;
        int result = INT_MAX;
        while (start <= end){
            int mid = start + (end - start)/2;
//            cout << start << " " << mid << " "<< end << " "<<isPossible(weights, D, mid)<< endl;
            if (isPossible(weights, D, mid)){
                result = min(result, mid);
                end = mid - 1;
            }
            else {
                start = mid + 1;
            }
        }
        return result;

    }
};
int main(){
    vector<int> arr = {3};
    Solution obj;
    cout << obj.shipWithinDays(arr, 3);

    return 0;
}
